pkg/distributor/distributor.go

```diff
@@ -892,10 +892,11 @@ func (d *Distributor) forwardSamples(ctx context.Context, userID string, ts []mi
 	}
 
 	endpoint := d.limits.ForwardingEndpoint(userID)
+	forwardingDropBefore := d.limits.ForwardingDropBefore(userID)
 
 	// Reassign req.Timeseries because the forwarder creates a new slice which has been filtered down.
 	// The cleanup func will cleanup the new slice, it's the forwarders responsibility to return the old one to the pool.
-	ts, forwardingErrCh = d.forwarder.Forward(ctx, endpoint, forwardingRules, ts)
+	ts, forwardingErrCh = d.forwarder.Forward(ctx, endpoint, forwardingDropBefore, forwardingRules, ts)
```

Comment:
I would suggest passing final timestamp (possibly zero) to Forward function, so that it doesn't need to do any conversion anymore. Now we have mix of methods taking duration and timestamps in this PR.

Reply:
good point, I'll do that

Reply:
I fixed it here 70feba2ed